File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        int res = 0;
        int l = 0;
        int chars[3];
        int charCount = 0;
        for(int r = 0; r < s.length(); r++)
        {
            chars[s[r] - 'a']++;
            if(chars[s[r] - 'a'] == 1)
            {
                charCount++;
            }
            while(charCount == 3)
            {
                res += (s.length() - r);
                chars[s[l] - 'a']--;
                if(chars[s[l] - 'a'] == 0)
                {
                    charCount--;
                }
                l++;
            }
        }
        return res;
    }
};

File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        res = 0
        l = 0
        chars = [0 for _ in range(3)]
        charCount = 0
        for r in range(len(s)):
            chars[ord(s[r]) - ord('a')] += 1
            if chars[ord(s[r]) - ord('a')] == 1:
                charCount += 1
            while charCount == 3:
                res += len(s) - r
                chars[ord(s[l]) - ord('a')] -= 1
                if chars[ord(s[l]) - ord('a')] == 0:
                    charCount -= 1
                l += 1
        return res